src/rdkafka_sasl_aws_msk_iam.c

```diff
@@ -0,0 +1,816 @@
+        str_builder_add_str(sb, "&");
+        str_builder_add_str(sb, "X-Amz-Algorithm=AWS4-HMAC-SHA256&");
+        str_builder_add_str(sb, "X-Amz-Credential=");
+        str_builder_add_str(sb, uri_credential);
+        str_builder_add_str(sb, "&");
+        str_builder_add_str(sb, "X-Amz-Date=");
+        str_builder_add_str(sb, uri_amz_date);
+        str_builder_add_str(sb, "&");
+        str_builder_add_str(sb, "X-Amz-Expires=900&");  // AWS recommends 900 seconds
+        
+        if (state->aws_security_token.ptr != NULL) {
+            char *uri_amz_security_token = rd_uri_encode(state->aws_security_token);
+            
+            str_builder_add_str(sb, "X-Amz-Security-Token=");
+            str_builder_add_str(sb, uri_amz_security_token);
+            str_builder_add_str(sb, "&");
+            
+            RD_IF_FREE(uri_amz_security_token, rd_free);
+        }
+        
+        str_builder_add_str(sb, "X-Amz-SignedHeaders=host");
+        
+        char *canonical_query_string = str_builder_dump(sb);
+        
+        str_builder_destroy(sb);
+        
+        RD_IF_FREE(uri_action, rd_free);
+        RD_IF_FREE(credential_wo_prefix.ptr, rd_free);
+        RD_IF_FREE(credential.ptr, rd_free);
+        RD_IF_FREE(uri_credential, rd_free);
+        RD_IF_FREE(amz_date.ptr, rd_free);
+        RD_IF_FREE(uri_amz_date, rd_free);
+        
+        return canonical_query_string;
+}
+
+
+/**
+ * @brief Generates a canonical request
+ * @remark canonical_request will be allocated and must be freed.
+ */
+static char *rd_kafka_sasl_aws_msk_iam_build_canonical_request (struct rd_kafka_sasl_aws_msk_iam_state *state) {
+        char *canonical_query_string = rd_kafka_sasl_aws_msk_iam_build_canonical_query_string(state);
+        const char *hostname = state->hostname.ptr;
+        int canonical_header_size = strlen(hostname) + strlen("host:\n") + 1;
+        char canonical_header_buf[256];
+        rd_snprintf(canonical_header_buf, canonical_header_size, "host:%s\n", hostname);
+        
+        const char *signed_headers = "host";
+        
+        unsigned char md_value[EVP_MAX_MD_SIZE];
+        unsigned int md_len, i;
+        EVP_MD_CTX *mdctx;
+        
+        mdctx = EVP_MD_CTX_new();
+        EVP_DigestInit_ex(mdctx, state->md, NULL);
+        EVP_DigestUpdate(mdctx, "", 0);  // payload should be empty string
+        EVP_DigestFinal_ex(mdctx, md_value, &md_len);
+        EVP_MD_CTX_free(mdctx);
+        
+        char res_hexstring[64];
+        for (i = 0; i < md_len; i++)
+               sprintf(&(res_hexstring[i * 2]), "%02x", md_value[i]);  // save string in hex base 16
+        
+        str_builder_t *sb;
+        sb = str_builder_create();
+        str_builder_add_str(sb, "GET\n");
+        str_builder_add_str(sb, "/\n");
+        str_builder_add_str(sb, canonical_query_string);
+        str_builder_add_str(sb, "\n");
+        str_builder_add_str(sb, canonical_header_buf);
```

Comment:
I ran into some unexpected behavior related to writing the `canonical_header_buf` here.
The value of `canonical_header_buf` looks good earlier in this block (verified in gdb) but I was losing it in the final `canonical_request`.

The `canonical_request` looked like this:

```
GET
/
Action=kafka-cluster%3AConnect&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=REDACTED%2F20210817%2Fus-east-1%2Fkafka-cluster%2Faws4_request&X-Amz-Date=20210817T184802Z&X-Amz-Expires=900&X-Amz-Security-Token=REDACTED&X-Amz-SignedHeaders=host

host
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Ignoring the `canonical_header_buf` and constructing the string like this fixed my issue:
```c
str_builder_add_str(sb, "\nhost:");
str_builder_add_str(sb, hostname);
str_builder_add_str(sb, "\n\n");
```

Reply:
removed the `canonical_header_buf`. not sure what was happening there but rather be safe on that one. 

I changed to `res_hexstring[65]` and made all of them consistent in terms of syntax and added the NUL terminator. since they are being directly copied into the string builder, it still worked with length 64 but for the sake of proper string creation in C, using 65 + the terminator is better.